### PR TITLE
Show error message when mouse over mark

### PR DESCRIPTION
**Pull request by matttheus**
To get this done I needed change the textarea to a div and add the attribute contenteditable="true" to the editor.

### CODE REVIEW

**Comment by matttheus** [comment]
image

**Comment by matttheus** [comment]
Related to #258.

File: website/public/demo/index.ejs.html [hunk 1]
```diff
@@ -121,14 +133,13 @@ <h2>Demo</h2>
       </p>
 
       <pre id="shadow-code-input"></pre>
-      <textarea id="code-input" spellcheck="false">
+<div contenteditable="true" id="code-input" spellcheck="false">
```
Comment (strager) [inline website/public/demo/index.ejs.html:136]: Blocking: contenteditable (and maybe some other things) seems to break a lot of things.

For example, if I clear the text box, then type 'a', the cursor stays *before* the 'a'. It should be *after* the 'a'. If I then type 's' then 'd' then 'f', the text box shows "fdsa" (instead of "asdf").

Weird things also happen when I type some digits then press the enter key a few times.

I'm pretty sure these issues are because of markEditorText. When I originally made the demo, I tried making contenteditable work, but there were too many issues to fight. For example, even after fixing the issues I mentioned above, undo-redo was broken.

Let's keep this as `<textarea>`, unless you can make the contenteditable approach high-quality. I don't think making contenteditable work is possible, though; even Gmail's email composer is full of bugs *in plain text mode*.

**Review by strager — CHANGES_REQUESTED** [review]
Some UI issues I noticed while playing around:

If I hover over an error, then type, the popup behaves oddly. Note the overlap:
Screen Shot 2021-05-20 at 23 41 44

If I then move my cursor away and back, shown errors don't match the code:
Screen Shot 2021-05-20 at 23 43 03
Screen Shot 2021-05-20 at 23 43 28

If an error message is really long, the message can easily go off screen:
Screen Shot 2021-05-20 at 23 45 27

**Review by strager — CHANGES_REQUESTED** [review]
Blocking: If I scroll the page, the error pop-ups show in the wrong place.

Overall, is looking good. Ping me if you hit issues moving code into the .mjs again.

**Comment by strager** [comment]
@matttheus Can you rebase/merge `master`? There's a conflict with `website/public/demo/index.ejs.html`.

**Comment by strager** [comment]
> > @matttheus Can you rebase/merge `master`? There's a conflict with `website/public/demo/index.ejs.html`.
> 
> Can you help solve this problem?
> 
> I already did `git pull origin master` and solved the conflicts. But I still getting this error.
> 
> ```shell
> [matheus@matheus-pc website]$ git push my github-pr/258 
> To github.com:matttheus/quick-lint-js.git
>  ! [rejected]        github-pr/258 -> github-pr/258 (non-fast-forward)
> error: failed to push some refs to 'github.com:matttheus/quick-lint-js.git'
> hint: Updates were rejected because the tip of your current branch is behind
> hint: its remote counterpart. Integrate the remote changes (e.g.
> hint: 'git pull ...') before pushing again.
> hint: See the 'Note about fast-forwards' in 'git push --help' for details.
> ```
> 
> ```shell
> # ~/.gitconfig
> [pull]
>     ff = only
>     rebase = true
> 
> [rebase]
>     autostash = true
> ```

It looks like you rebased your public commits. If you do that, and you want to update your public branch with your rewritten commits, you need to force-push.

Easy way:

    $ git push --force-with-lease my github-pr/258

File: website/public/demo/demo.mjs [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 import { createProcessFactoryAsync } from "quick-lint-js-wasm/quick-lint-js.js";
 import { markEditorText } from "./editor.mjs";
+// import { showErrorMessageBox } from "./error-message.mjs";
```
Comment (strager) [inline website/public/demo/demo.mjs:6]: Delete commented-out code.

File: website/public/demo/demo.mjs [hunk 3]
```diff
@@ -26,6 +27,7 @@ createProcessFactoryAsync()
 
       // TODO(strager): On crash, show the error to the user.
       let input = codeInputElement.value;
+
```
Comment (strager) [inline website/public/demo/demo.mjs:30]: Nit: Undo unrelated whitespace change.

File: website/public/demo/editor.mjs [hunk 4]
```diff
@@ -88,6 +88,7 @@ class EditorMarker {
       this._currentMarkIndex < this._marks.length
         ? this._marks[this._currentMarkIndex]
         : null;
+
```
Comment (strager) [inline website/public/demo/editor.mjs:91]: Nit: Undo unrelated whitespace change.

File: website/public/demo/index.ejs.html [hunk 5]
```diff
@@ -51,6 +52,35 @@
...
+        text-decoration-style: wavy;
+      }
+      mark:empty::before {
+        color: red;
+        content: url("carat.svg") / "^";
+        content: url("carat.svg");
+        margin-left: -2px; /* (width of SVG) / 2 */
+        margin-top: 4px; /* height of SVG */
+        position: absolute;
+      }
```
Comment (strager) [inline website/public/demo/index.ejs.html:82]: This is mostly copy-pasted from main.css. Remove this copy.

File: website/public/demo/error-message.mjs [hunk 6]
```diff
@@ -0,0 +1,57 @@
+function showErrorMessageBox(mark) {
```
Comment (strager) [inline website/public/demo/error-message.mjs:1]: Shouldn't all this code be inside demo.mjs? Why a separate file?

File: website/public/demo/error-message.mjs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  const shadowInput = document.querySelector("#shadow-code-input");
+  Array.from(shadowInput.querySelectorAll("mark")).every((mark) => {
+    const markRect = mark.getBoundingClientRect();
+    if (cursorOverMark(event.clientX, event.clientY, markRect)) {
+      showErrorMessageBox(mark);
+      return false;
+    }
+    removeErrorMessageBox();
+    return true;
+  });
```
Comment (strager) [inline website/public/demo/error-message.mjs:43]: Are you using `every` so you can exit the loop early? A `for` loop seems much more appropriate:
```suggestion
  for (let mark of shadowInput.querySelectorAll("mark")) {
    const markRect = mark.getBoundingClientRect();
    if (cursorOverMark(event.clientX, event.clientY, markRect)) {
      showErrorMessageBox(mark);
      break;
    }
    removeErrorMessageBox();
  }
```

File: website/public/demo/error-message.mjs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+function showErrorMessage(event) {
+  const shadowInput = document.querySelector("#shadow-code-input");
+  Array.from(shadowInput.querySelectorAll("mark")).every((mark) => {
+    const markRect = mark.getBoundingClientRect();
+    if (cursorOverMark(event.clientX, event.clientY, markRect)) {
+      showErrorMessageBox(mark);
+      return false;
+    }
+    removeErrorMessageBox();
```
Comment (strager) [inline website/public/demo/error-message.mjs:41]: Why are we calling `removeErrorMessageBox` every iteration? Could we call it once before looping instead?

Comment (strager) [inline website/public/demo/error-message.mjs:41]: Blocking: We should call `removeErrorMessageBox` at least once. If there are no marks, then `removeErrorMessageBox` isn't called. This causes the error message box to stick around with no way to hide it.

File: website/public/demo/error-message.mjs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    top: ${Math.trunc(bottom)}px; 
+    left: ${Math.trunc(left)}px`
+  );
+  div.classList.add("error-box");
+  return div;
+}
+
+function removeErrorMessageBox() {
+  const errorMessagesBoxs = document.querySelectorAll("#error-message-box");
+  errorMessagesBoxs.forEach((box) => box.remove());
```
Comment (strager) [inline website/public/demo/error-message.mjs:30]: If there can be multiple elements with ID error-message-box, then we should use a class or a JS array instead of an ID.

If there can only be zero or one elements with ID error-message-box, then I think it's a bit confusing to loop here.

File: website/public/demo/error-message.mjs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  let div = document.createElement("div");
+  const { bottom, left, height } = markedElement.getBoundingClientRect();
+  div.setAttribute("id", "error-message-box");
+  div.innerText = `${code} - ${errorMessage}`;
+  div.setAttribute(
+    "style",
+    `position: fixed; 
+    top: ${Math.trunc(bottom)}px; 
+    left: ${Math.trunc(left)}px`
+  );
```
Comment (strager) [inline website/public/demo/error-message.mjs:23]: Nit: Isn't it more "clean" to use `div.style.position = 'fixed';`, etc.? I'm not up to date with the hottest ways to set inline styles, so maybe what you have is recommended.

File: website/public/demo/error-message.mjs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    mark.attributes["data-severity"].value
+  );
+  let body = document.querySelector("body");
+  body.appendChild(div);
+}
+
+function createErrorBox(markedElement, errorMessage, code, severity) {
+  // TODO: Change background color based of the severity
+  let div = document.createElement("div");
+  const { bottom, left, height } = markedElement.getBoundingClientRect();
```
Comment (strager) [inline website/public/demo/error-message.mjs:15]: Nit: The `height` variable is unused.
```suggestion
  const { bottom, left } = markedElement.getBoundingClientRect();
```

File: website/public/demo/error-message.mjs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  const leftRightIn =
+    cursorPosX >= markRect.left && cursorPosX <= markRect.left + markRect.width;
+  return topDownIn && leftRightIn;
+}
+
+document.addEventListener("DOMContentLoaded", () => {
+  document
+    .querySelector("#code-input")
+    .addEventListener("mousemove", showErrorMessage);
+});
```
Comment (strager) [inline website/public/demo/error-message.mjs:57]: We should also update the error message box after the user types. Otherwise, we might have a stale error message box corresponding to no error.

File: website/src/config.mjs [hunk 13]
```diff
@@ -13,6 +13,10 @@ export let websiteConfig = {
       entryPoints: ["/demo/demo.mjs"],
       external: ["fs", "path"],
     },
+    "/demo/error-message.bundled.js": {
```
Comment (strager) [inline website/src/config.mjs:16]: Blocking: We should reuse the existing bundle instead of making a new bundle.

**Review by strager — CHANGES_REQUESTED** [review]
Nit: Run `yarn fmt` in the `website` directory to automatically format your code.

**Review by strager — APPROVED** [review]
Bug: If I move the mouse cursor over an error, then type on the keyboard to fix the error, the error box is still showing. I expect the box to disappear after fixing the error.

**Comment by strager** [comment]
> Bug: If I move the mouse cursor over an error, then type on the keyboard to fix the error, the error box is still showing. I expect the box to disappear after fixing the error.

VS Code hides error message boxes when you type, even if the error is still present. Perhaps we should implement VS Code's behavior.

**Comment by matttheus** [comment]
> Bug: If I move the mouse cursor over an error, then type on the keyboard to fix the error, the error box is still showing. I expect the box to disappear after fixing the error.

I fixed it adding a event listener to the textarea to remove the error message box and it changes. I did the same to the click event on the textarea.

**Comment by matttheus** [comment]
> > Bug: If I move the mouse cursor over an error, then type on the keyboard to fix the error, the error box is still showing. I expect the box to disappear after fixing the error.
> 
> VS Code hides error message boxes when you type, even if the error is still present. Perhaps we should implement VS Code's behavior.

Yeah, I added a onchange event listener.

File: website/public/demo/demo.mjs [hunk 14]
```diff
@@ -55,6 +55,83 @@ function synchronizeSize() {
...
+  removeErrorMessageBox();
+
+  const shadowInput = document.querySelector("#shadow-code-input");
+  const marks = shadowInput.querySelectorAll("mark");
+  for (let i = 0; i < marks.length; i++) {
+    const mark = marks[i];
+    const markRect = mark.getBoundingClientRect();
+    if (
+      cursorOverMark(event.clientX, event.clientY, markRect) &&
+      !errorMessageBoxAlreadyExist()
```
[the remark this answers is not in the document]
Comment (matttheus) [reply website/public/demo/demo.mjs]: > When would an error message box ever exist here? We removed all of them on line 102, and line 113 runs at most once per call to `showErrorMessage`.

I remove this checking

File: website/public/demo/demo.mjs [hunk 15]
```diff
@@ -55,6 +55,83 @@ function synchronizeSize() {
...
+  div.style.left = `${posCursorX}px`;
+  div.classList.add("error-box");
+  return div;
+}
+
+function removeErrorMessageBox() {
+  const errorMessagesBoxs = document.querySelectorAll(".error-message-box");
+  for (let i = 0; i < errorMessagesBoxs.length; i++) {
+    errorMessagesBoxs[i].remove();
+  }
```
[the remark this answers is not in the document]
Comment (matttheus) [reply website/public/demo/demo.mjs]: > Nit: Would a for-of loop work here?

I remove the loop, now it remove a single error message box.

File: website/public/demo/demo.mjs [hunk 16]
```diff
@@ -55,6 +55,83 @@ function synchronizeSize() {
...
+  // TODO: Change background color based of the severity
+  let div = document.createElement("div");
+  const { bottom } = markedElement.getBoundingClientRect();
+  div.setAttribute("class", "error-message-box");
+  div.innerText = `${code} - ${errorMessage}`;
+  div.style.position = "fixed";
+  div.style.overflow = "auto";
+  div.style.top = `${Math.trunc(bottom)}px`;
+  div.style.left = `${posCursorX}px`;
+  div.classList.add("error-box");
```
[the remark this answers is not in the document]
Comment (matttheus) [reply website/public/demo/demo.mjs]: > Did you mean to add two different classes (error-message-box and error-box)? I see that JS uses one and CSS uses the other. Why not have one class?

Was a mistake, I changed the reference to this div to be id, instead of classes.

File: website/public/demo/demo.mjs [hunk 17]
```diff
@@ -55,6 +55,71 @@ function synchronizeSize() {
...
+  document.querySelector("#error-box")?.remove();
+}
+
+function showErrorMessage(event) {
+  removeErrorMessageBox();
+
+  const shadowInput = document.querySelector("#shadow-code-input");
+  const marks = shadowInput.querySelectorAll("mark");
+  for (let i = 0; i < marks.length; i++) {
+    const mark = marks[i];
```
Comment (strager) [inline website/public/demo/demo.mjs:99]: Nit: Would a for-of loop work?
```suggestion
  const marks = shadowInput.querySelectorAll("mark");
  for (let mark of marks) {
```

**Review by strager — APPROVED** [review]

File: website/public/demo/demo.mjs (hunk 17 above)
Comment (matttheus) [reply website/public/demo/demo.mjs]: > Nit: Would a for-of loop work?

Yeah, worked. I fixed.

**Comment by strager** [comment]
Looks like a test is failing:
```
Running markEditorText mark first word on line ...
error properties: Object({ generatedMessage: true, code: 'ERR_ASSERTION', actual: '<mark data-message="undefined" data-code="undefined" data-severity="undefined">hello</mark> world', expected: '<mark>hello</mark> world', operator: 'strictEqual' })
AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected
+ '<mark data-message="undefined" data-code="undefined" data-severity="undefined">hello</mark> world'
- '<mark>hello</mark> world'
    at UserContext.<anonymous> (file:///__w/quick-lint-js/quick-lint-js/website/test/test-demo.mjs:20:12)
    at <Jasmine>
    at processImmediate (internal/timers.js:461:21)
FAIL    markEditorText mark first word on line
```
Fix it so I can merge this PR.

**Comment by matttheus** [comment]
> Looks like a test is failing:
> 
> ```
> Running markEditorText mark first word on line ...
> error properties: Object({ generatedMessage: true, code: 'ERR_ASSERTION', actual: '<mark data-message="undefined" data-code="undefined" data-severity="undefined">hello</mark> world', expected: '<mark>hello</mark> world', operator: 'strictEqual' })
> AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
> + actual - expected
> + '<mark data-message="undefined" data-code="undefined" data-severity="undefined">hello</mark> world'
> - '<mark>hello</mark> world'
>     at UserContext.<anonymous> (file:///__w/quick-lint-js/quick-lint-js/website/test/test-demo.mjs:20:12)
>     at <Jasmine>
>     at processImmediate (internal/timers.js:461:21)
> FAIL    markEditorText mark first word on line
> ```
> 
> Fix it so I can merge this PR.

I fixed it, but I am getting this error:

```bash
Running error documentation html wraps <mark>-d byte order mark ...
Error: Expected '<article id=''><p>code:</p>
<figure><pre><code><mark data-message="undefined" data-code="undefined" data-severity="undefined"></mark>--BOM
</code></pre></figure></article>' to contain '<code><mark data-message="undefined" data-code="undefined" data-severity="undefined"><span class='unicode-bom'></span></mark>--BOM'.
    at <Jasmine>
    at UserContext.<anonymous> (file:///home/matheus/Dev/personal/quick-lint-js/website/test/test-error-documentation.mjs:156:26)
    at <Jasmine>
FAIL    error documentation html wraps <mark>-d byte order mark
error Command failed with exit code 4.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Seems to be this tests it is failing at `test-error-documentation.mjs`.
```js
  it("html wraps <mark>-d byte order mark", () => {
    let doc = ErrorDocumentation.parseString(
      "file.md",
      "code:\n\n    \ufeff--BOM\n"
    );
    doc.diagnostics = [[{ begin: 0, end: 1 }]];
    expect(doc.toHTML()).toContain(
      "<code><mark><span class='unicode-bom'>\u{feff}</span></mark>--BOM"
    );
  });
```


This is what I tried.
```js
  it("html wraps <mark>-d byte order mark", () => {
    let doc = ErrorDocumentation.parseString(
      "file.md",
      "code:\n\n    \ufeff--BOM\n"
    );
    doc.diagnostics = [[{ begin: 0, end: 1 }]];
    expect(doc.toHTML()).toContain(
      `<code><mark data-message="undefined" data-code="undefined" data-severity="undefined"><span class='unicode-bom'>\u{feff}</span></mark>--BOM`
    );
  });
```
